common/src/main/java/io/pravega/common/concurrent/ScheduledQueue.java

```diff
@@ -0,0 +1,345 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.common.concurrent;
+
+import com.google.common.collect.Iterators;
+import java.util.AbstractQueue;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nonnull;
+import lombok.Data;
+import lombok.val;
+
+/**
+ * Provides a unbounded blocking queue which {@link Scheduled} items can be added to.
+ * Items which are scheduled will not be returned from {@link #poll()} or {@link #take()} until their scheduled time.
+ */
+public class ScheduledQueue<E extends Scheduled> extends AbstractQueue<E> implements BlockingQueue<E> {
+   
+    private final AtomicLong itemsAdded = new AtomicLong(0);
+    private final AtomicLong itemsRemoved = new AtomicLong(0);
+    private final ConcurrentSkipListMap<FireTime, E> delayedTasks = new ConcurrentSkipListMap<FireTime, E>();
+    private final ConcurrentLinkedQueue<E> readyTasks = new ConcurrentLinkedQueue<E>();
+    private final Semaphore blocker = new Semaphore(1);
+    
+    @Data
+    private static final class FireTime implements Comparable<FireTime> {
+        private final long timeNanos;
+        private final long sequenceNumber;
+
+        @Override
+        public int compareTo(FireTime other) {
+            if (this.timeNanos < other.timeNanos) {
+                return -1;
+            } else if (this.timeNanos > other.timeNanos) {
+                return 1;
+            } else {
+                return Long.compare(this.sequenceNumber, other.sequenceNumber);
+            }
+        }
+    }
+
+    /**
+     * Retrieves and removes the head of this queue, waiting if necessary
+     * until an element becomes available.
+     *
+     * @return the head of this queue
+     * @throws InterruptedException if interrupted while waiting
+     */
+    @Override
+    public E take() throws InterruptedException {
+        return poll(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+    }
+
+
+    @Override
+    public E poll(long timeout, TimeUnit unit) throws InterruptedException {
+        blocker.tryAcquire();
```

Comment:
If the thread is interrupted it will throw as soon as it attempts to block below. We have no need to check for iterupted early. It's very unlikely as it should only occur upon shutdown, so it doesn't provide an advantage to avoid one map lookup in that case.